m3/tests/reporter_test.cc

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <chrono>
+#include <cstdint>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <thread>
+
+#include "gtest/gtest.h"
+#include "thrift/protocol/TCompactProtocol.h"
+#include "thrift/server/TSimpleServer.h"
+#include "thrift/transport/TBufferTransports.h"
+#include "thrift/transport/TServerSocket.h"
+
+#include "m3/reporter.h"
+#include "m3/reporter_builder.h"
+#include "m3/thrift/M3.h"
+#include "m3/thrift/m3_types.h"
+#include "m3/udp_transport.h"
+#include "mock_handler.h"
+#include "mock_server.h"
+
+class ReporterTest : public ::testing::Test {
```

Comment:
mind adding a test to ensure the destructor logic is good too

Reply:
Not sure what you had in mind?

Reply:
something like: create a scope and reporter that report every 10s(or 0s if that stops bg flushing), create a counter and inc it, and then ensure the scope+reporter lifecycles end before the 10s are over; and the metric is still flushed by the destructor

Reply:
The reporter and scope tests should cover their own lifecycles I think. I can add an integration test for the scope and reporter together though.

Reply:
Sounds good.